File: codewars/authoring/test/codewars/AISOTest.java
package codewars;

import org.junit.Test;

import java.lang.reflect.Constructor;
import java.util.Arrays;
import java.util.Optional;
import java.util.function.Function;

import static codewars.ISO.*;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

/**
 * Created by ice1000 on 17-6-19.
 *
 * @author ice1000
 */
public class ISOTest {
	private static <A, B> ISO<A, B> iso(Function<A, B> forward, Function<B, A> backward) {
		return new ISO<A, B>() {
			@Override
			public A bw(B b) {
				return backward.apply(b);
			}

			@Override
			public B fw(A a) {
				return forward.apply(a);
			}
		};
	}

	private static final String MEOW = "meow";
	private static final String MOE = "moe";

	private static ISO<Boolean, String> tISO = iso(it -> it ? MEOW : MOE, MEOW::equals);

	private static <A, B> Function<A, A> lrl(ISO<A, B> iso) {
		return it -> iso.bw(iso.fw(it));
	}

	private static <A, B> Function<B, B> rlr(ISO<A, B> iso) {
		return it -> iso.fw(iso.bw(it));
	}

	private static String cul(Object a, Object b) {
		return a + " " + b;
	}

	private static String cur(Tuple<?, ?> a) {
		return a.a + " " + a.b;
	}

	private static ISO<Either<Either<Integer, Boolean>, String>, Either<Integer, Either<Boolean, String>>> pa() {
		return plusAssoc();
	}

	private static ISO<Tuple<Tuple<Integer, Boolean>, String>, Tuple<Integer, Tuple<Boolean, String>>> ma() {
		return multAssoc();
	}

	private static ISO<Tuple<Integer, Either<Boolean, String>>, Either<Tuple<Integer, Boolean>, Tuple<Integer, String>>> di() {
		return dist();
	}

	private static ISO<Function<Integer, Function<Boolean, String>>, Function<Tuple<Integer, Boolean>, String>> cu() {
		return curryISO();
	}

	private static final String str = "it was me, DIO!";

	@Test
	public void subStLTest() {
		assertEquals(MEOW, subStL(tISO).apply(true));
		assertEquals(false, subStL(isoBool()).apply(false));
		assertEquals(true, subStL(isoBool()).apply(true));
	}

	@Test
	public void subStRTest() {
		assertEquals(MEOW, subStL(subStR(iso(null, e -> tISO)).apply(null)).apply(true));
		assertEquals(true, subStR(isoBool()).apply(true));
		assertEquals(false, subStR(isoBool()).apply(false));
	}

	@Test
	@SuppressWarnings("unchecked")
	public void isoSymmTest() {
		assertEquals(MOE, subStR(subStR(isoSymm()).apply((ISO) tISO)).apply(false));
	}

	@Test
	public void isoStreamTest() {
		assertEquals(Optional.of(MEOW + MEOW + MOE),
				subStL(isoStream(tISO)).apply(Arrays.stream(new Boolean[]{true, true, false}))
						.reduce((a, b) -> a + b));
	}

	@Test
	public void isoFuncTest() {
		assertEquals(MEOW, subStL(isoFunc(symm(tISO), tISO)).apply(MEOW::equals).apply(true));
	}

	@Test
	@SuppressWarnings({"unchecked", "AssertEqualsBetweenInconvertibleTypes"})
	public void reflTest() {
		assertEquals(MEOW, subStL((ISO<Boolean, Boolean>) subStL(refl()).apply(tISO)).apply(true));
	}

	@Test
	public void symmTest() {
		assertEquals(MEOW, subStR(symm(tISO)).apply(true));
	}

	@Test
	public void transTest() {
		assertEquals(MEOW, subStL(trans(refl(), trans(tISO, refl()))).apply(true));
	}

	private static void assertTupleEquals(Tuple<?, ?> a, Tuple<?, ?> b) {
		if (a.a instanceof Either) assertEitherEquals((Either<?, ?>) a.a, (Either<?, ?>) b.a);
		else if (a.a instanceof Tuple) assertTupleEquals((Tuple<?, ?>) a.a, (Tuple<?, ?>) b.a);
		else assertEquals(a.a, b.a);
		if (a.b instanceof Either) assertEitherEquals((Either<?, ?>) a.b, (Either<?, ?>) b.b);
		else if (a.b instanceof Tuple) assertTupleEquals((Tuple<?, ?>) a.b, (Tuple<?, ?>) b.b);
		else assertEquals(a.b, b.b);
	}

	private static void assertEitherEquals(Either<?, ?> a, Either<?, ?> b) {
		if (a.l instanceof Tuple) assertTupleEquals((Tuple<?, ?>) a.l, (Tuple<?, ?>) b.l);
		else if (a.l instanceof Either) assertEitherEquals((Either<?, ?>) a.l, (Either<?, ?>) b.l);
		else /*if (null != a.l)*/ assertEquals(a.l, b.l);
		if (a.r instanceof Tuple) assertTupleEquals((Tuple<?, ?>) a.r, (Tuple<?, ?>) b.r);
		else if (a.r instanceof Either) assertEitherEquals((Either<?, ?>) a.r, (Either<?, ?>) b.r);
		else /*if (null != a.r)*/ assertEquals(a.r, b.r);
	}

	@SuppressWarnings({"OptionalUsedAsFieldOrParameterType", "ConstantConditions"})
	private static void assertOptionalEquals(Optional<?> x, Optional<?> y) {
		if (!x.isPresent() && !y.isPresent()) return;
		Object a = x.get();
		Object b = x.get();
		if (a instanceof Tuple) assertTupleEquals((Tuple<?, ?>) a, (Tuple<?, ?>) a);
		else if (a instanceof Either) assertEitherEquals((Either<?, ?>) a, (Either<?, ?>) b);
		else assertEquals(a, b);
	}

	/**
	 * 秘技！类型擦除！
	 *
	 * @return multSO' in Haskell
	 */
	@SuppressWarnings("unchecked")
	private static ISO<Tuple<Unit, Object>, Object> multSO_() {
		return trans(isoProd(one(), (ISO) refl()), trans((ISO) multS(), trans(isoPlus(refl(), multO()), trans(plusComm(), plusO()))));
	}

	private static <B> ISO<Either<Unit, B>, Optional<B>> plusSO_() {
		return trans(isoPlus(one(), refl()), trans(plusS(), isoS(plusO())));
	}

	private static ISO<Either<Optional<String>, Integer>, Optional<Either<String, Integer>>> pss() {
		return plusS();
	}

	private static ISO<Either<Unit, Integer>, Optional<Integer>> pso() {
		return plusSO_();
	}

	private static ISO<Tuple<Optional<String>, Integer>, Either<Integer, Tuple<String, Integer>>> mss() {
		return multS();
	}

	private static ISO<Tuple<Unit, Object>, Object> mso() {
		return multSO_();
	}

	private static ISO<Function<Optional<Integer>, String>, Tuple<String, Function<Integer, String>>> poss() {
		return powS();
	}

	/**
	 * 秘技！类型擦除！
	 *
	 * @return powSO' in Haskell
	 */
	@SuppressWarnings("unchecked")
	private static ISO<Function<Unit, Object>, Object> powSO_() {
		return trans((ISO) isoPow(one(), (ISO) refl()), trans((ISO) powS(),
				trans(isoProd(refl(), powO()), trans(multComm(), (ISO) multSO_()))));
	}

	/** master spark! */
	private static String func(Integer integer) {
		String[] strings = new String[integer];
		Arrays.fill(strings, "DA★ZE");
		return String.join("|", strings);
	}

	@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
	private static String mfunc(Optional<Integer> optional) {
		return optional.map(ISOTest::func).orElse(str);
	}

	private static ISO<Function<Unit, String>, String> poso() {
		return powSO();
	}

	private static ISO<Function<Unit, Object>, Object> poso_() {
		return powSO_();
	}

	@Test
	public void assoc() {
		assertEitherEquals(Either.left(Either.left(233)), lrl(pa()).apply(Either.left(Either.left(233))));
		assertEitherEquals(Either.left(Either.right(true)), lrl(pa()).apply(Either.left(Either.right(true))));
		assertEitherEquals(Either.right(str), lrl(pa()).apply(Either.right(str)));
		assertEitherEquals(Either.left(233), rlr(pa()).apply(Either.left(233)));
		assertEitherEquals(Either.right(Either.left(true)), rlr(pa()).apply(Either.right(Either.left(true))));
		assertEitherEquals(Either.right(Either.right(str)), rlr(pa()).apply(Either.right(Either.right(str))));
	}

	@Test
	public void distTest() {
		Tuple<Integer, Either<Boolean, String>> tuple = new Tuple<>(0, Either.left(true));
		assertTupleEquals(tuple, lrl(di()).apply(tuple));
		Tuple<Integer, Either<Boolean, String>> tuple2 = new Tuple<>(0, Either.right(str));
		assertTupleEquals(tuple2, lrl(di()).apply(tuple2));
		Either<Tuple<Integer, Boolean>, Tuple<Integer, String>> either = Either.left(new Tuple<>(0, true));
		assertEitherEquals(either, rlr(di()).apply(either));
		Either<Tuple<Integer, Boolean>, Tuple<Integer, String>> either2 = Either.right(new Tuple<>(0, str));
		assertEitherEquals(either2, rlr(di()).apply(either2));
	}

	@Test
	public void curryTest() {
		assertEquals(cul(0, true), lrl(cu()).apply(a -> b -> cul(a, b)).apply(0).apply(true));
		assertEquals(cul(0, true), rlr(cu()).apply(ISOTest::cur).apply(new Tuple<>(0, true)));
	}

	@Test
	public void oneTest() {
		assertEquals(Unit.INSTANCE, lrl(one()).apply(Unit.INSTANCE));
		assertEquals(Optional.empty(), rlr(one()).apply(Optional.empty()));
	}

	@Test
	public void twoTest() {
		assertTrue(lrl(two()).apply(true));
		assertFalse(lrl(two()).apply(false));
		assertEquals(Optional.empty(), rlr(two()).apply(Optional.empty()));
		assertEquals(Optional.of(Optional.empty()), rlr(two()).apply(Optional.of(Optional.empty())));
	}

	@Test
	public void plusTest() {
		assertEitherEquals(Either.right(str), lrl(plusO()).apply(Either.right(str)));
		assertEquals(str, rlr(plusO()).apply(str));
		assertEitherEquals(Either.left(Optional.of(str)), lrl(pss()).apply(Either.left(Optional.of(str))));
		assertEitherEquals(Either.left(Optional.empty()), lrl(pss()).apply(Either.left(Optional.empty())));
		assertEitherEquals(Either.right(233), lrl(pss()).apply(Either.right(233)));
		assertOptionalEquals(Optional.of(Either.left(str)), rlr(pss()).apply(Optional.of(Either.left(str))));
		assertOptionalEquals(Optional.of(Either.right(233)), rlr(pss()).apply(Optional.of(Either.right(233))));
		assertOptionalEquals(Optional.empty(), rlr(pss()).apply(Optional.empty()));
		assertEitherEquals(Either.left(Unit.INSTANCE), lrl(pso()).apply(Either.left(Unit.INSTANCE)));
		assertEitherEquals(Either.right(233), lrl(pso()).apply(Either.right(233)));
		assertOptionalEquals(Optional.empty(), rlr(pso()).apply(Optional.empty()));
		assertOptionalEquals(Optional.of(233), rlr(pso()).apply(Optional.of(233)));
	}

	@Test
	public void multTest() {
		assertTupleEquals(new Tuple<>(Optional.of(str), 233), lrl(mss()).apply(new Tuple<>(Optional.of(str), 233)));
		assertTupleEquals(new Tuple<>(Optional.empty(), 233), lrl(mss()).apply(new Tuple<>(Optional.empty(), 233)));
		assertEitherEquals(Either.left(233), rlr(mss()).apply(Either.left(233)));
		assertEitherEquals(Either.right(new Tuple<>(str, 233)), rlr(mss()).apply(Either.right(new Tuple<>(str, 233))));
		assertTupleEquals((new Tuple<>(Unit.INSTANCE, true)), lrl(mso()).apply((new Tuple<>(Unit.INSTANCE, true))));
		assertTupleEquals((new Tuple<>(Unit.INSTANCE, false)), lrl(mso()).apply((new Tuple<>(Unit.INSTANCE, false))));
		assertTrue((Boolean) rlr(mso()).apply(true));
		assertFalse((Boolean) rlr(mso()).apply(false));
	}

	@Test
	public void powTest() {
		assertEquals(Unit.INSTANCE, rlr(powO()).apply(Unit.INSTANCE));
		assertEquals("DA★ZE|DA★ZE|DA★ZE", lrl(poss()).apply(ISOTest::mfunc).apply(Optional.of(3)));
		assertEquals(str, lrl(poss()).apply(ISOTest::mfunc).apply(Optional.empty()));
		assertEquals(str, rlr(poss()).apply(new Tuple<>(str, ISOTest::func)).a);
		assertEquals("DA★ZE|DA★ZE|DA★ZE", rlr(poss()).apply(new Tuple<>(str, ISOTest::func)).b.apply(3));
		assertEquals(str, lrl(poso()).apply(unit -> str).apply(Unit.INSTANCE));
		assertEquals(str, rlr(poso()).apply(str));
		assertEquals(str, lrl(poso_()).apply(unit -> str).apply(Unit.INSTANCE));
		assertEquals(str, rlr(poso_()).apply(str));
	}

	@Test
	public void privateConstructors() {
		for (Constructor<?> constructor : Unit.class.getConstructors()) {
			assertFalse(constructor.isAccessible());
		}
		for (Constructor<?> constructor : Either.class.getConstructors()) {
			assertFalse(constructor.isAccessible());
		}
	}
}

File: codewars/authoring/kotlin/AISOTest.kt
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Test
import java.util.*
import java.util.Arrays.stream
import java.util.Optional.empty
import java.util.Optional.of
import java.util.stream.Collectors
import java.util.stream.Stream
import kotlin.test.*


class ISOTest {

	@Test
	fun subStLTest() {
		assertEquals(MEOW, subStL(tISO)(true))
		assertEquals(false, subStL(isoBool())(false))
		assertEquals(true, subStL(isoBool())(true))
	}

	@Test
	fun subStRTest() {
		assertEquals(MEOW, subStL(subStR(iso({ _: ISO<Boolean, String> -> Any() }, { _: Any -> tISO }))(Any()))(true))
		assertEquals(true, subStR(isoBool())(true))
		assertEquals(false, subStR(isoBool())(false))
	}

	@Test
	fun reflTest() = assertEquals(MEOW, subStL(subStL(refl<ISO<Boolean, String>>())(tISO))(true))

	@Test
	fun symmTest() = assertEquals(MEOW, subStR(symm(tISO))(true))

	@Test
	fun transTest() = assertEquals(MEOW, subStL(trans(refl(), trans(tISO, refl())))(true))

	@Test
	fun isoPairTest() = assertEquals(MEOW to MOE, subStL(isoPair(tISO, tISO))(Pair(true, false)))

	@Test
	fun isoStreamTest() = assertEquals(of(MEOW + MEOW + MOE), subStL(isoStream(tISO))(stream(arrayOf(true, true, false))).reduce(String::plus))

	@Test
	fun isoFuncTest() = assertEquals(MEOW, subStL(isoFunc(symm(tISO), tISO))(MEOW::equals)(true))

	@Test
	fun isoSymmTest() = assertEquals(MOE, subStR(subStR(isoSymm<String, Boolean>())(tISO))(false))

	@Test
	fun isoOptionalTest() {
		assertFalse(subStL(isoOptional(tISO))(empty()).isPresent)
		assertEquals(MEOW, subStL(isoOptional(tISO))(of(true)).get())
	}

	@Test
	fun isoUnOptionalTest() {
		assertEquals(false, subStL(bISO)(true))
		assertEquals(true, subStL(bISO)(false))
		assertEquals(false, subStL(badISO)(true))
		assertFailsWith<NoSuchElementException> { subStL(badISO)(false) }
		val bi = isoUnOptional(isoOptional(refl<Any>()))
		assertEquals(true, lrl(bi)(true))
		assertEquals(false, lrl(bi)(false))
		assertEquals(true, rlr(bi)(true))
		assertEquals(false, rlr(bi)(false))
		val biso = symm(isoUnOptional(symm(mISO)))
		assertEquals(false, subStL(bISO)(true))
		assertEquals(true, subStL(bISO)(false))
		assertEquals(false, subStL(badISO)(true))
		assertFailsWith<NoSuchElementException> { subStL(badISO)(false) }
	}

	private fun lu(): Stream<Unit> = Stream.of(Unit, Unit, Unit, Unit, Unit, Unit, Unit, Unit)

	@Test
	fun isoEUTest() {
		assertEquals(conv(lrl(isoEU())(left(Stream.empty<Unit>()))), conv(left<Stream<Unit>, Nothing>(Stream.empty())))
		assertEquals(conv(lrl(isoEU())(left(lu()))), conv(left<Stream<Unit>, Nothing>(lu())))
		assertEquals(conv(lrl(isoEU())(right(Unit))), conv(right(Unit)))
		assertEquals(conv(rlr(isoEU())(left(Stream.empty<Unit>()))), conv(left<Stream<Unit>, Nothing>(Stream.empty())))
		assertEquals(conv(rlr(isoEU())(left(lu()))), conv(left<Stream<Unit>, Nothing>(lu())))
		assertEquals(subStL(isoEU())(right(Unit)).pm({ true }, { false }), true)
	}

	@Test
	fun isoEitherTest() {
		assertEquals(left<Boolean, Nothing>(false), subStL(isoEither(isoBoolNot(), tISO))(left(true)))
		assertEquals(left<Boolean, Nothing>(true), subStL(isoEither(isoBoolNot(), tISO))(left(false)))
		assertEquals(right<Boolean, String>(MEOW), subStL(isoEither(isoBoolNot(), tISO))(right(true)))
	}

	@Test
	fun assoc() {
		assertEquals(left<Either<Int, Boolean>, String>(left(233)), lrl(pa())(left(left(233))))
		assertEquals(left<Either<Int, Boolean>, String>(right(true)), lrl(pa())(left(right(true))))
		assertEquals(right<Either<Int, Boolean>, String>(str), lrl(pa())(right(str)))
		assertEquals(left<Int, Either<Boolean, String>>(233), rlr(pa())(left(233)))
		assertEquals(right<Int, Either<Boolean, String>>(left(true)), rlr(pa())(right(left(true))))
		assertEquals(right<Int, Either<Boolean, String>>(right(str)), rlr(pa())(right(right(str))))
	}

	@Test
	fun distTest() {
		val tuple = Pair(0, left<Boolean, String>(true))
		assertEquals(tuple, lrl(di())(tuple))
		val tuple2 = Pair(0, right<Boolean, String>(str))
		assertEquals(tuple2, lrl(di())(tuple2))
		val either = left<Pair<Int, Boolean>, Pair<Int, String>>(Pair(0, true))
		assertEquals(either, rlr(di())(either))
		val either2 = right<Pair<Int, Boolean>, Pair<Int, String>>(Pair(0, str))
		assertEquals(either2, rlr(di())(either2))
	}

	@Test
	fun curryTest() {
		assertEquals(cul(0, true), lrl(cu())({ a -> { b -> cul(a, b) } })(0)(true))
		assertEquals(cul(0, true), rlr(cu())(CO::cur)(Pair(0, true)))
	}

	@Test
	fun oneTest() {
		assertEquals(Unit, lrl(one())(Unit))
		assertEquals(Optional.empty<Optional<Unit>>(), rlr(one())(Optional.empty()))
	}

	@Test
	fun twoTest() {
		assertTrue(lrl(two())(true))
		assertFalse(lrl(two())(false))
		assertEquals(Optional.empty<Optional<Unit>>(), rlr(two())(Optional.empty()))
		assertEquals(Optional.of<Optional<Unit>>(Optional.empty()), rlr(two())(Optional.of(Optional.empty())))
	}

	@Test
	fun plusTest() {
		assertEquals(right<Boolean, String>(str), lrl(plusO<String>())(right(str)))
		assertEquals(str, rlr(plusO<String>())(str))
		assertEquals(left<Optional<String>, Nothing>(Optional.of(str)), lrl(pss())(left(Optional.of(str))))
		assertEquals(left<Optional<String>, Nothing>(Optional.empty()), lrl(pss())(left(Optional.empty())))
		assertEquals(right<Optional<String>, Int>(233), lrl(pss())(right(233)))
		assertEquals(Optional.of(left<String, Int>(str)), rlr(pss())(Optional.of(left(str))))
		assertEquals(Optional.of(right<String, Int>(233)), rlr(pss())(Optional.of(right(233))))
		assertEquals(Optional.empty<Nothing>(), rlr(pss())(Optional.empty()))
		assertEquals(left<Unit, Int>(Unit), lrl(pso())(left(Unit)))
		assertEquals(right<Unit, Int>(233), lrl(pso())(right(233)))
		assertEquals(Optional.empty<Nothing>(), rlr(pso())(Optional.empty()))
		assertEquals(Optional.of(233), rlr(pso())(Optional.of(233)))
	}

	@Test
	fun multTest() {
		assertEquals(Optional.of(str) to 233, lrl(mss())(Optional.of(str) to 233))
		assertEquals(Optional.empty<String>() to 233, lrl(mss())(Optional.empty<String>() to 233))
		assertEquals(left<Int, Pair<String, Int>>(233), rlr(mss())(left(233)))
		assertEquals(right<Int, Pair<String, Int>>(str to 233), rlr(mss())(right(str to 233)))
		assertEquals(Unit to true, lrl(mso())(Unit to true))
		assertEquals(Unit to false, lrl(mso())(Unit to false))
		assertTrue(rlr(mso())(true))
		assertFalse(rlr(mso())(false))
	}

	@Test
	fun powTest() {
		assertEquals(Unit, rlr(powO<Nothing>())(Unit))
		assertEquals("DA★ZE|DA★ZE|DA★ZE", lrl(poss())(CO::mfunc)(Optional.of(3)))
		assertEquals(str, lrl(poss())(CO::mfunc)(Optional.empty()))
		assertEquals(str, rlr(poss())(Pair(str, CO::func)).first)
		assertEquals("DA★ZE|DA★ZE|DA★ZE", rlr(poss())(Pair(str, CO::func)).second(3))
		assertEquals(str, lrl(poso())({ str })(Unit))
		assertEquals(str, rlr(poso())(str))
		assertEquals(str, lrl(poso_())({ str })(Unit))
		assertEquals(str, rlr(poso_())(str))
	}

	@Test
	fun shouldntHackByChangingEquals() {
		assertNotEquals(right(Pair(str, 666)), rlr(mss())(right(Pair(str, 233))))
		assertNotEquals(right(Pair("str", 233)), rlr(mss())(right(Pair(str, 233))))
	}

	companion object CO {
		private inline fun <A, B> iso(crossinline forward: (A) -> B, crossinline backward: (B) -> A): ISO<A, B> = object : ISO<A, B>() {
			override fun bw(b: B): A = backward(b)
			override fun fw(a: A): B = forward(a)
		}

		private fun cul(a: Any, b: Any): String = "$a $b"
		private fun cur(a: Pair<Any, Any>): String = "${a.first} ${a.second}"
		private fun pa(): ISO<Either<Either<Int, Boolean>, String>, Either<Int, Either<Boolean, String>>> = plusAssoc()
		private fun ma(): ISO<Pair<Pair<Int, Boolean>, String>, Pair<Int, Pair<Boolean, String>>> = multAssoc()
		private fun di(): ISO<Pair<Int, Either<Boolean, String>>, Either<Pair<Int, Boolean>, Pair<Int, String>>> = dist()
		private fun cu(): ISO<(Int) -> (Boolean) -> String, (Pair<Int, Boolean>) -> String> = curryISO()
		private fun <B> plusSO_(): ISO<Either<Unit, B>, Optional<B>> = trans(isoPlus(one(), refl<B>()), trans(plusS(), isoS(plusO<B>())))
		private fun <B> multSO_(): ISO<Pair<Unit, B>, B> = trans(isoProd(one(), refl<B>()), trans(multS(), trans(isoPlus(refl<B>(), multO<B>()), trans(plusComm(), plusO()))))
		private fun mfunc(optional: Optional<Int>): String = optional.map(CO::func).orElse(str)
		private fun pso(): ISO<Either<Unit, Int>, Optional<Int>> = plusSO_()
		private fun mss(): ISO<Pair<Optional<String>, Int>, Either<Int, Pair<String, Int>>> = multS()
		private fun mso(): ISO<Pair<Unit, Boolean>, Boolean> = multSO_()
		private fun poss(): ISO<(Optional<Int>) -> String, Pair<String, (Int) -> String>> = powS()
		private fun pss(): ISO<Either<Optional<String>, Int>, Optional<Either<String, Int>>> = plusS()
		private fun poso(): ISO<(Unit) -> String, String> = powSO()
		private fun poso_(): ISO<(Unit) -> String, String> = powSO_()
		private fun <A> powSO_(): ISO<(Unit) -> A, A> =
				trans(isoPow(one(), refl<A>()), trans(powS(), trans(isoProd(refl<A>(), powO<A>()), trans(multComm(), multSO()))))

		/** master spark! */
		private fun func(integer: Int): String {
			val strings = arrayOfNulls<String>(integer)
			Arrays.fill(strings, "DA★ZE")
			return strings.joinToString("|")
		}

		private val MEOW = "meow"
		private val MOE = "moe"
		private val tISO = iso({ it: Boolean -> if (it) MEOW else MOE }, { it: String -> MEOW == it })
		private val mISO = iso<Optional<Boolean>, Optional<Boolean>>({ it.map({ if (it) of(false) else empty() }).orElseGet({ of(true) }) }, { it })
		private val bISO = isoUnOptional(mISO)
		private val badMISO = iso<Optional<Boolean>, Optional<Boolean>>({ it.flatMap({ if (it) of(false) else empty() }) }, { it })
		private val badISO = isoUnOptional(badMISO)
		private fun <A, B> lrl(iso: ISO<A, B>): (A) -> A = { iso.bw(iso.fw(it)) }
		private fun <A, B> rlr(iso: ISO<A, B>): (B) -> B = { iso.fw(iso.bw(it)) }
		private val str = "it was me, DIO!"
		private val lu = stream(arrayOf(0, 1, 2, 3, 4, 5, 6, 7))
		internal fun <A> conv(e: Either<Stream<Unit>, A>): Either<List<Unit>, A> = e.pm({ left(it.collect(Collectors.toList())) }, ::right)
	}
}